sql/sql_parse.cc

```diff
@@ -7861,7 +7861,17 @@ static bool wsrep_mysql_parse(THD *thd, char *rawbuf, uint length,
                       DBUG_ASSERT(!debug_sync_set_action(thd, STRING_WITH_LEN(act)));
                     });
 #endif
-        WSREP_DEBUG("wsrep retrying AC query: %lu  %s",
+        /* The flag is needed to detect retried CTAS to add OR REPLACE clause
+         * to protect against potentially existing table left from the
+         * previously failed execution. */
+        if (thd->lex->sql_command == SQLCOM_CREATE_TABLE &&
```

Comment:
This unfortunately looks a bit unsafe. The `mysql_parse()` which is called from this function sets up the lex tree and may also clean it up. I'm not sure if it currently happens in MariaDB, but there have been bugs in MySQL which were caused by relying on `thd->lex` contents in this outer scope.

Even if this works right now, it may break if the implementation inside `mysql_parse()` changes in the future.

How important it is to keep the CTAS retrying functionality? The BF abort should not be too common (and BF abort happening might even be a bug), and this retry logic seems to complicate things quite a lot.

Reply:
Keeping or not keeping CTAS retrying is not a correct question. It is retried because of the retry logic. The question is how we can determine CTAS to either:
- not retry it
- retry it properly with adding OR REPLACE clause.

If the approach is not correct, I will try to look for the other one.

Reply:
Implemented CTAS retry by setting retry counter to threshold value.